File: c++/20040.cpp
#include <iostream>

using namespace std;

int arr[1000000][2];
int parent[500000];

int getParent(int x) {
  if (parent[x] == x)
    return x;
  else
    return parent[x] = getParent(parent[x]);
}

void unionParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  if (a < b)
    parent[b] = a;
  else
    parent[a] = b;
}

bool sameParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  return a == b;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int n, m;
  cin >> n >> m;

  for (int i = 0; i < m; i++) {
    cin >> arr[i][0] >> arr[i][1];
  }
  for (int i = 0; i < n; i++) {
    parent[i] = i;
  }

  for (int i = 0; i < m; i++) {
    int a = arr[i][0];
    int b = arr[i][1];

    if (!sameParent(a, b)) {
      unionParent(a, b);
    } else {
      cout << i + 1;
      return 0;
    }
  }

  cout << 0;

  return 0;
}

File: js/20040.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((arr) => arr.split(' ').map(Number));

const [n, m] = inputs.shift();

const parent = Array(n)
  .fill()
  .map((_, i) => i);

console.log(solve());

function solve() {
  for (let i = 0; i < m; i++) {
    const [a, b] = inputs[i];
    if (!sameParent(a, b)) {
      unionParent(a, b);
    } else {
      return i + 1;
    }
  }

  return 0;
}

function getParent(k) {
  if (parent[k] === k) return k;
  else return (parent[k] = getParent(parent[k]));
}

function unionParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
}

function sameParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  return a === b;
}
